src/main/java/com/devcycle/sdk/server/local/managers/EventQueueManager.java

```diff
@@ -0,0 +1,174 @@
+            public void run() {
+                try {
+                    flushEvents();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        };
+
+        scheduler.scheduleAtFixedRate(getConfigRunnable, 0, this.eventFlushIntervalMS, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Flush events in queue to DevCycle Events API. Requeue events if flush fails
+     */
+    public void flushEvents() throws Exception {
+        if (isFlushingEvents) return;
+
+        if (serverKey == null || serverKey.equals("")) {
+            throw new Exception("DevCycle is not yet initialized to publish events.");
+        }
+
+        FlushPayload[] flushPayloads = new FlushPayload[0];
+        try {
+            flushPayloads = this.localBucketing.flushEventQueue(this.serverKey);
+        } catch (Exception e) {
+            System.out.printf("DVC Error Flushing Events: %s%n", e.getMessage());
+        }
+
+        if (flushPayloads.length == 0) return;
+        System.out.printf("AS Flush Payloads: %s%n", Arrays.toString(flushPayloads));
+
+        int eventCount = 0;
+        isFlushingEvents = true;
+        for (FlushPayload payload: flushPayloads) {
+            eventCount += payload.eventCount;
+            publishEvents(this.serverKey, payload);
+        }
+        isFlushingEvents = false;
+        System.out.printf("DVC Flush %d AS Events, for %d Users%n", eventCount, flushPayloads.length);
+    }
+
+    /**
+     * Queue DVCAPIEvent for publishing to DevCycle Events API.
+     */
+    public void queueEvent(User user, Event event) throws Exception {
+        if (checkEventQueueSize()) {
+            System.out.printf("Max event queue size reached, dropping event: %s%n", event);
+            return;
+        }
+
+        this.localBucketing.queueEvent(this.serverKey, OBJECT_MAPPER.writeValueAsString(user), OBJECT_MAPPER.writeValueAsString(event));
+    }
+
+    /**
+     * Queue DVCEvent that can be aggregated together, where multiple calls are aggregated
+     * by incrementing the 'value' field.
+     */
+    public void queueAggregateEvent(Event event, BucketedUserConfig bucketedConfig) throws Exception {
+        if (checkEventQueueSize()) {
+            System.out.printf("Max event queue size reached, dropping aggregate event: %s%n", event);
+            return;
+        }
+
+        if (bucketedConfig != null) {
+            this.localBucketing.queueAggregateEvent(this.serverKey, OBJECT_MAPPER.writeValueAsString(event), OBJECT_MAPPER.writeValueAsString(bucketedConfig.variableVariationMap));
+        } else {
+            this.localBucketing.queueAggregateEvent(this.serverKey, OBJECT_MAPPER.writeValueAsString(event), "{}");
+        }
+    }
+
+    private void publishEvents(String serverKey, FlushPayload flushPayload) throws InterruptedException {
+        Thread publishEventsThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
```

Comment:
Does this callback on a different thread? or the same as the calling thread? just worried about threading calling WASM.

Reply:
Hm, not sure if I got what you're asking completely but I think when we do `publishEventsThread.start()` it creates a new thread which then does `run()` on that newly created thread
https://www.geeksforgeeks.org/difference-between-thread-start-and-thread-run-in-java/

Reply:
@kaushalkapasi do we have any thread locking around the calls to WASM?

Reply:
no, we currently don't... this is the only area where we are use threads to interact with WASM for the `onPayloadSuccess / onPayloadFailure` callbacks

Reply:
Ok it should be ok... but this is potentially risky, we should probably only call WASM from the same thread.